From 6d11373d0c4e8aa8cbfb362b8324fafa65864fcf Mon Sep 17 00:00:00 2001
From: t <tristan75015@gmail.com>
Date: Mon, 12 Oct 2026 02:48:00 +0200
Subject: [PATCH] MFront DDIF2 brick: checks of the options

- only one of softening_slope, softening_slopes, fracture_energy and
  fracture_energies can be given: the check only failed when all four
  were given, the other combinations failing later with "name 'Rp'
  already registred";
- when the first component of fracture_stresses, softening_slopes or
  fracture_energies is not constant, the sign of the constant components
  following it is checked too (it was only checked when the first
  component is constant).
---
 mfront/src/DDIF2StressPotential.cxx | 21 +++++++++++++++++----
 1 file changed, 17 insertions(+), 4 deletions(-)

diff --git a/mfront/src/DDIF2StressPotential.cxx b/mfront/src/DDIF2StressPotential.cxx
index c08916b30..a54b71c00 100644
--- a/mfront/src/DDIF2StressPotential.cxx
+++ b/mfront/src/DDIF2StressPotential.cxx
@@ -100,6 +100,18 @@ namespace mfront::bbrick {
         }
         addParameter(bd, vn, en, 3, values);
       } else {
+        // the constant components following a non constant one are
+        // checked too
+        for (unsigned short i = 1; i != 3; ++i) {
+          if (!mps[i].is<ConstantMaterialProperty>()) {
+            continue;
+          }
+          if (b) {
+            checkIsStrictlyPositive(mps[i]);
+          } else {
+            checkIsStrictlyNegative(mps[i]);
+          }
+        }
         addLocalVariable(bd, vt, vn, 3);
       }
     }
@@ -198,10 +210,11 @@ namespace mfront::bbrick {
     // data
     throw_if((d.count("fracture_stress")) && (d.count("fracture_stresses")),
              "can't specify 'fracture_stress' and 'fracture_stresses");
-    throw_if((d.count("softening_slope")) && (d.count("softening_slopes")) &&
-                 (d.count("fracture_energy")) && (d.count("fracture_energies")),
-             "only of of 'softening_slope', 'softening_slopes', "
-             "'fracture_energy' and 'fracture_energies' must be specified");
+    throw_if(d.count("softening_slope") + d.count("softening_slopes") +
+                     d.count("fracture_energy") + d.count("fracture_energies") >
+                 1,
+             "only one of 'softening_slope', 'softening_slopes', "
+             "'fracture_energy' and 'fracture_energies' can be specified");
     extractDDIF2MaterialProperty(this->sr, bd, dsl, "fracture_stress",
                                  "fracture_stresses", "sigr", "stress",
                                  "FractureStress", d, true);
-- 
2.43.0

